core(emulation): refactor emulation settings & CLI flags (#11779)
GoogleChrome/lighthouse: docs/emulation.md

# Emulation in Lighthouse

In Lighthouse, "Emulation" refers to the screen/viewport emulation and UserAgent string spoofing.

> ["Throttling"](./throttling.md) covers the similar topics around network connection throttling/simulation and CPU throttling.

Reviewer: ```suggestion
["Throttling"](./throttling.md) covers the similar topics around network and CPU throttling/simulation.
```

felt simpler 🤷

> In Lighthouse v7, most of the configuration regarding emulation changed to be more intuitive and clear. The [tracking issue](https://github.com/GoogleChrome/lighthouse/issues/10910

Reviewer: there's something about this sentence being *in the PR that brings those updates* being really funny to me 😆 

I love it. Feels odd as the 2nd sentence for people just stumbling onto the generic emulation docs though, move down to "Changes in v7"?

> With the default configuration, Lighthouse emulates a mobile device. There's [a `desktop` configuration](../lighthouse-core/config/desktop-config.js), available to CLI users with `--preset=desktop`, which applies a consistent desktop environment and scoring calibration. This is recommended as a replacement for `--emulation-form-factor=desktop`.

Reviewer: ```suggestion
With the default configuration, Lighthouse emulates a mobile device. There's [a `desktop` configuration](../lighthouse-core/config/desktop-config.js), available to CLI users with `--preset=desktop`, which applies a consistent desktop environment and scoring calibration. This is recommended as a replacement for `--form-factor=desktop`.
```

Or if this was referring to the v6 flags then `--emulated-form-factor`?

Author: ah yeah typo. i meant to refer to refer to that (i added this to docs excepting we'll get people who used -eff=desktop without problems before and want the quick answer)

thx

### Advanced emulation setups

> Some products use Lighthouse in scenarios where emulation is applied outside of Lighthouse (eg by Puppeteer) or running against Chrome on real mobile devices.

Reviewer: ```suggestion
Some products use Lighthouse in scenarios where emulation is applied outside of Lighthouse (e.g. by Puppeteer) or running against Chrome on real mobile devices.
```

### Advanced emulation setups

Some products use Lighthouse in scenarios where emulation is applied outside of Lighthouse (eg by Puppeteer) or running against Chrome on real mobile devices.

You must always set `formFactor`. It doesn't control emulation, but it determines how Lighthouse should interpret the run in regards to scoring performance metrics and skipping mobile-only tests in desktop.

You can choose how `screenEmulation` is applied. It can accept an object of `{width: number, height: number, deviceScaleRatio: number, mobile: boolean}` to apply that screen emulation or `false` if Lighthouse should avoid applying screen emulation. It's typically set to `false` if either emulation is applied outside of Lighthouse, or it's being run on a mobile device. The `mobile` boolean applies overlay scrollbars and a few other mobile-specific screen emulation characteristics.

> You can choose how to handle userAgent emulation. The `emulatedUserAgent` property accepts either a `string` to apply the provided userAgent or `false` if no UA spoofing should be applied. Typically `false` is used if UA spoofing is apply outside of Lighthouse or on a mobile device. You can also redundantly apply userAgent emulation with no risk.

Reviewer: ```suggestion
You can choose how to handle userAgent emulation. The `emulatedUserAgent` property accepts either a `string` to apply the provided userAgent or `false` if no UA spoofing should be applied. Typically `false` is used if UA spoofing is applied outside of Lighthouse or on a mobile device. You can also redundantly apply userAgent emulation with no risk.
```

> If you're using Lighthouse against a mobile device, you want to set `--no-screenEmulation` and `--throttling.cpuSlowdownMultiplier=1`. (`--formFactor=mobile` is the default already).

Reviewer: ```suggestion
If you're using Lighthouse on a mobile device, you want to set `--no-screenEmulation` and `--throttling.cpuSlowdownMultiplier=1`. (`--formFactor=mobile` is the default already).
```

or `running .... against`?

Reviewer: and does `--no-screen-emulation` work?

In Lighthouse v7, most of the configuration regarding emulation changed to be more intuitive and clear. The [tracking issue](https://github.com/GoogleChrome/lighthouse/issues/10910
) captures additional motivations.

* Removed: The `emulatedFormFactor` property (which determined how emulation is applied).

> * Removed: The `TestedAsMobileDevice` artifact. Instead of being inferred, the  `formFactor` property is used.
> * Removed: The `internalDisableDeviceScreenEmulation` property. It's equivalent to the new `screenEmulation: false`.

Reviewer: ```suggestion
* Removed: The `TestedAsMobileDevice` artifact. Instead of being inferred, the explicit `formFactor` property is used.
* Removed: The `internalDisableDeviceScreenEmulation` property. It's equivalent to the new `--screenEmulation.disabled=true`.
```

I think this got buried because you fixed adam's while my review was in progress.

> * Removed: The `TestedAsMobileDevice` artifact is removed. Instead of being inferred, the  `formFactor` property is used.
> * The `internalDisableDeviceScreenEmulation` property is removed. It's equivalent to the new `screenEmulation: false`.

Reviewer 2: ```suggestion
* Removed: The `TestedAsMobileDevice` artifact. Instead of being inferred, the  `formFactor` property is used.
* Removed: The `internalDisableDeviceScreenEmulation` property. It's equivalent to the new `screenEmulation: false`.
```
nit

Reviewer: I'll raise you one more :)

```suggestion
* Removed: The `TestedAsMobileDevice` artifact. Instead of being inferred, the explicit `formFactor` property is used.
* Removed: The `internalDisableDeviceScreenEmulation` property. It's equivalent to the new `--screenEmulation.disabled=false`.
```

as for the `Removed` styling I have no strong preference for what happens but definitely agree with adam that something consistent would be preferable
